keps/sig-node/20190131-runas-groupid.md

```diff
@@ -14,21 +13,266 @@ approvers:
+  SupplementalGroups SupplementalGroupsStrategyOptions
+
+
+  // RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+  RunAsGroup RunAsGroupStrategyOptions
+   .....
+}
+
+// RunAsGroupStrategyOptions defines the strategy type and any options used to create the strategy.
+ type RunAsUserStrategyOptions struct {
+     // Rule is the strategy that will dictate the allowable RunAsGroup values that may be set.
+     Rule RunAsGroupStrategy
+     // Ranges are the allowed ranges of gids that may be used.
+     // +optional
+     Ranges []GroupIDRange
+ }
+
+// RunAsGroupStrategy denotes strategy types for generating RunAsGroup values for a
+ // SecurityContext.
+ type RunAsGroupStrategy string
+ 
+ const (
+     // container must run as a particular gid.
+     RunAsGroupStrategyMustRunAs RunAsGroupStrategy = "MustRunAs"
+     // container may make requests for any gid.
+     RunAsGroupStrategyRunAsAny RunAsGroupStrategy = "RunAsAny"
+ )
+```
+
+## Behavior
+
+Following points should be noted:
+
+- `FSGroup` and `SupplementalGroups` will continue to have their old meanings and would be untouched.  
+- The `RunAsGroup` In the SecurityContext will override the `RunAsGroup` in the PodSecurityContext.
+- If both `RunAsUser` and `RunAsGroup` are NOT provided, the USER field in Dockerfile is used
+- If both `RunAsUser` and `RunAsGroup` are specified, that is passed directly as User.
+- If only one of `RunAsUser` or `RunAsGroup` is specified, the remaining value is decided by the Runtime,
+  where the Runtime behavior is to make it run with uid or gid as 0.
+
+Basically, we guarantee to set the values provided by user, and the runtime dictates the rest.
+
+Here is an example of what gets passed to docker User
+- runAsUser set to 9999, runAsGroup set to 9999 -> Config.User set to 9999:9999
+- runAsUser set to 9999, runAsGroup unset -> Config.User set to 9999 -> docker runs you with 9999:0
+- runAsUser unset, runAsGroup set to 9999 -> Config.User set to :9999 -> docker runs you with 0:9999 
+- runAsUser unset, runAsGroup unset -> Config.User set to whatever is present in Dockerfile
+This is to keep the behavior backward compatible and as expected.
+
+### Note About RunAsNonRoot field
+
+Note that this change does not introduce an equivalent field called runAsNonRootGroup in both SecurityContext
+and PodSecurityContext. There was ongoing discussion about this field at PR [#62216](https://github.com/kubernetes/kubernetes/pull/62217)
+The summary of this discussion seems as follows:-
+- Use PSP MustRunAs Group strategy to guarantee that Pod never runs with 0 as Primary Group ID.
+- Using the PSP MustRunaAs Group strategy forces Pod to always specify a RunAsGroup
+- RunAsGroup field when specified in PodSpec, will always override USER field in Dockerfile
+
+There are other potentially unresolved discussions in that PR which need a followup.
+
+## Summary of Changes needed
```

Comment:
I think you can remove this section. It's likely to end up out of date very quickly. If you want, you can link to the PRs that implemented these changes instead.

Reply:
done @tallclair

Reply:
done